### PR TITLE
kv: remove timeouts

### PR DESCRIPTION
**Pull request by tamird**
Reviewer note: this is nearly all code deletion when viewed with whitespace ignored, except for one line in `pkg/kv/transport.go` where I hardcoded `time.Second` in the clients' deadline field.

~~I don't feel great about removing `TestAmbiguousCommitDueToLeadershipChange`, but I don't know how to rewrite that test in the absence of `SendNextTimeout`; the scenario tested there seems no longer to be possible.~~

### CODE REVIEW

**Comment by petermattis** [comment]
The `SendNextTimeout` removal looks ok. I don't recall the history of `PendingRPCTimeout`. Cc @spencerkimball 

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-:-Kon7zonigx4mrhEThaU:bz7td57)*

**Comment by bdarnell** [comment]
Reviewed 5 of 12 files at r2, 7 of 7 files at r3.
Review status: 11 of 12 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/kv/transport.go, line 304 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KorNkbUmFIeOV82CsQW:-KorNkbUmFIeOV82CsQX:b-hcog9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/b238bb52542dc9c167b92f3567d960b5113e8d3a/pkg/kv/transport.go#L304)):*
> ```Go
> 			gt.orderedClients[i].retryable = retryable
> 			if retryable {
> 				gt.orderedClients[i].deadline = timeutil.Now().Add(time.Second)
> ```

I think we can remove this deadline and the "resurrect retryables" logic too - it was effectively disabled anyway with SendNextTimeout set to a huge value. 

---

*[pkg/sql/ambiguous_commit_test.go, line 50 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KorPr6Hc25G0qRIS3JT:-KorPr6Hc25G0qRIS3JU:bj9b08m) ([raw file](https://github.com/cockroachdb/cockroach/blob/7177c64eb8cf74b41741185c0df6d5ac576ae5c6/pkg/sql/ambiguous_commit_test.go#L50)):*
> ```Go
> // advance, it can result in violated uniqueness constraints, or
> // duplicate key violations. See #6053, #7604, and #10023.
> func TestAmbiguousCommitDueToLeadershipChange(t *testing.T) {
> ```

I think this test is still relevant. To test it without SendNextTimeout, we need a kv.Transport that returns a GRPC error (with a code other than Unavailable - it should use whichever code gets returned when RPC heartbeats fail and the connection gets closed). And because of this, the last commit removes too much. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987)*

File: pkg/kv/dist_sender.go [hunk 1]
```diff
@@ -471,6 +450,7 @@ func (ds *DistSender) sendSingleRange(
 
 	br, err := ds.sendRPC(ctx, desc.RangeID, replicas, ba)
 	if err != nil {
+		log.ErrEvent(ctx, err.Error())
```
Comment (spencerkimball) [inline pkg/kv/dist_sender.go]: Was this for debugging, or is this error just not otherwise captured in the trace further up the call stack?

**Comment by tamird** [comment]
Review status: 11 of 12 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/kv/transport.go, line 304 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KorNkbUmFIeOV82CsQW:-Kor_UWij-phBwJy2Fy6:bk0ldvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b238bb52542dc9c167b92f3567d960b5113e8d3a/pkg/kv/transport.go#L304)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think we can remove this deadline and the "resurrect retryables" logic too - it was effectively disabled anyway with SendNextTimeout set to a huge value. 
</blockquote></details>

I agree that it was effectively disabled, but don't we want to keep it? I think disabling it was an accident.

---

*[pkg/sql/ambiguous_commit_test.go, line 50 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KorPr6Hc25G0qRIS3JT:-Kor_ZhfURg8ayUz25bv:byvihvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/7177c64eb8cf74b41741185c0df6d5ac576ae5c6/pkg/sql/ambiguous_commit_test.go#L50)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think this test is still relevant. To test it without SendNextTimeout, we need a kv.Transport that returns a GRPC error (with a code other than Unavailable - it should use whichever code gets returned when RPC heartbeats fail and the connection gets closed). And because of this, the last commit removes too much. 
</blockquote></details>

Agreed; will work on restoring this test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987)*

File: pkg/kv/dist_sender.go [hunk 2]
```diff
@@ -1158,51 +1135,54 @@ func (ds *DistSender) sendToReplicas(
...
-
-	// Send the first request.
-	pending := 1
-	if log.V(2) || log.HasSpanOrEvent(ctx) {
-		log.VEventf(ctx, 2, "r%d: sending batch %s to %s",
-			rangeID, args.Summary(), transport.NextReplica())
-	}
+	// Must be unbuffered because tests have blocking SendNext
+	// implementations.
+	done := make(chan BatchCall, 1)
```
Comment (spencerkimball) [inline pkg/kv/dist_sender.go]: The comment says the channel is unbuffered, but it has a buffer length of 1.

File: pkg/kv/dist_sender.go [hunk 3]
```diff
@@ -1158,51 +1135,54 @@ func (ds *DistSender) sendToReplicas(
...
+	log.VEventf(ctx, 2, "r%d: sending batch %s to %s", rangeID, args.Summary(), transport.NextReplica())
```
Comment (spencerkimball) [inline pkg/kv/dist_sender.go]: I don't know offhand what `log.HasSpanOrEvent(ctx)` actually does; I assume you do and this change is fully intentional.

File: pkg/kv/dist_sender.go [hunk 4]
```diff
@@ -1158,51 +1135,54 @@ func (ds *DistSender) sendToReplicas(
...
+				// lost, then the batch will succeed and we can be assured the
+				// commit was applied just once.
+				//
+				// The Unavailable code is used by GRPC to indicate that a
+				// request fails fast and is not sent, so we can be sure there
+				// is no ambiguity on these errors. Note that these are common
+				// if a node is down.
+				// See https://github.com/grpc/grpc-go/blob/52f6504dc290bd928a8139ba94e3ab32ed9a6273/call.go#L182
+				// See https://github.com/grpc/grpc-go/blob/52f6504dc290bd928a8139ba94e3ab32ed9a6273/stream.go#L158
+				if haveCommit && grpc.Code(err) != codes.Unavailable {
```
Comment (spencerkimball) [inline pkg/kv/dist_sender.go]: This change is not correct. Previously, if we got any RPC error, we continued to try the other replicas, hoping one would succeed. Even if we get an RPC error other than `Unavailable`, if we succeed with the batch at any additional replica, we can still return success instead of the `AmbiguousResultError`. Instead, we're now just fast-failing with the `AmbiguousResultError`.

This is one of the cases that all of this code was trying to solve for. We know that node1 is the leaseholder, we send node1 the batch, it times out because it's behind a network partition, and we retry to a node2 and node3, one of which will be elected Raft leader and gain the lease. Assuming that the original batch didn't succeed at node1, then it will succeed at the new leaseholder (either node2 or node3).

File: pkg/kv/dist_sender.go [hunk 5]
```diff
@@ -1158,51 +1135,54 @@ func (ds *DistSender) sendToReplicas(
...
+					return nil, roachpb.NewAmbiguousResultError(fmt.Sprintf("error=%s", err))
+				}
+				log.ErrEventf(ctx, "RPC error: %s", err)
+			} else {
 				// Determine whether the error must be propagated immediately or whether
```
Comment (spencerkimball) [inline pkg/kv/dist_sender.go]: This comment should probably be adjusted to reflect the fact there may not actually be an error. It's confusing as is.

File: pkg/kv/dist_sender.go [hunk 6]
```diff
@@ -1235,115 +1214,29 @@ func (ds *DistSender) sendToReplicas(
...
-					// return an ambiguous commit error instead of the
-					// returned error.
-					if haveCommit {
-						if pending > 0 || ambiguousError != nil {
-							log.ErrEventf(ctx, "returning ambiguous result (error=%v pending=%d)", ambiguousError, pending)
-							return nil, roachpb.NewAmbiguousResultError(
-								fmt.Sprintf("error=%v pending=%d", ambiguousError, pending))
-						}
-					}
+					// replicas.
```
Comment (spencerkimball) [inline pkg/kv/dist_sender.go]: Here, you're going to need to resurrect the `ambiguousError` variable and check `haveCommit` to properly return the `AmbiguousResultError` instead of whatever proximate error is received here.

File: pkg/sql/ambiguous_commit_test.go [hunk 7]
```diff
@@ -1,183 +0,0 @@
...
-	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
-	"github.com/cockroachdb/cockroach/pkg/storage"
-	"github.com/cockroachdb/cockroach/pkg/testutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
-	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
-	"github.com/lib/pq"
-)
-
-// TestAmbiguousCommitDueToLeadershipChange verifies that an ambiguous
```
Comment (spencerkimball) [inline pkg/sql/ambiguous_commit_test.go]: Definitely do not remove this test! This circumstance still is vitally important to solve for (it's the condition behind our original failure on the Jepsen tests).

File: pkg/kv/dist_sender.go [hunk 8]
```diff
@@ -1235,115 +1214,29 @@ func (ds *DistSender) sendToReplicas(
...
 					return call.Reply, nil
 				}
 
-				// Extract the detail so it can be included in the error
-				// message if this is our last replica.
-				//
+				log.ErrEventf(ctx, "application error: %s", call.Reply.Error)
+			}
+
+			if transport.IsExhausted() {
```
Comment (spencerkimball) [inline pkg/kv/dist_sender.go]: Now that you've removed the `SendNextTimeout()`, nothing uses the `deadline` being set in the `grpcTransport`. This means that in the case where node2 and node3 believe node1 (AWOL) is still the leaseholder, node1 will keep getting moved to the front of the transport, will fail fast, then we'll try node2 and node3 in rapid succession, each putting node1 back into the front position.

The end result is that because you're no longer employing deadline to trigger a timer, this code will busy loop, sending RPCs to node2 and node3 until a Raft election and leaseholder transfer to node2 or node3.

**Comment by spencerkimball** [comment]
Review status: 11 of 12 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/kv/transport.go, line 304 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KorNkbUmFIeOV82CsQW:-Kori2UmkJzLbY0opSJJ:b83n7ig) ([raw file](https://github.com/cockroachdb/cockroach/blob/b238bb52542dc9c167b92f3567d960b5113e8d3a/pkg/kv/transport.go#L304)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I agree that it was effectively disabled, but don't we want to keep it? I think disabling it was an accident.
</blockquote></details>

This unfortunately was an overlooked bug in our setting the `SendNextTimeout` value to 10 minutes. The case where it was incorrectly no longer used is as follows:

- Node1 is leaseholder and goes AWOL
- gRPC heartbeats indicate node1 is not reachable, and fail the RPC
- Node2 is tried and returns `Node1 is leaseholder`
- Node1 is put to front of transport
- Batch is sent to Node1 and fails fast
- Node3 is tried and returns `Node1 is leaseholder`
- Node1 is put to front of transport
- Batch is sent to Node1 and fails fast

Note that each time node2 and node3 fail with a `NotLeaseholderError`, they get the retryable flag set in the transport, which made them eligible for a retry. We do want to retry them because we expect either node2 or node3 to eventually secure the lease.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987)*

**Comment by tamird** [comment]
Review status: 11 of 12 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/kv/dist_sender.go, line 453 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-453:-KorjCe-SLq07OT1IwES:b-dzkxt4) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L453)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Was this for debugging, or is this error just not otherwise captured in the trace further up the call stack?
</blockquote></details>

The various errors returned here were previously being `ErrEvent`ed in the body of `sendToReplicas` - moving it here is just DRYer.

---

*[pkg/kv/dist_sender.go, line 1140 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1140:-KorjLDyeg0UPP6XUioh:b-wjy86e) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1140)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

The comment says the channel is unbuffered, but it has a buffer length of 1.
</blockquote></details>

Yup, should say "buffered". Done.

---

*[pkg/kv/dist_sender.go, line 1141 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1141:-KorjQcbgHID7lF0hs1T:b-jbqurj) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1141)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I don't know offhand what `log.HasSpanOrEvent(ctx)` actually does; I assume you do and this change is fully intentional.
</blockquote></details>

Yup, old code and new code is equivalent.

---

*[pkg/kv/dist_sender.go, line 1181 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1181:-KorkkMa1NixM1YzZi96:b-xokkgz) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1181)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

This change is not correct. Previously, if we got any RPC error, we continued to try the other replicas, hoping one would succeed. Even if we get an RPC error other than `Unavailable`, if we succeed with the batch at any additional replica, we can still return success instead of the `AmbiguousResultError`. Instead, we're now just fast-failing with the `AmbiguousResultError`.

This is one of the cases that all of this code was trying to solve for. We know that node1 is the leaseholder, we send node1 the batch, it times out because it's behind a network partition, and we retry to a node2 and node3, one of which will be elected Raft leader and gain the lease. Assuming that the original batch didn't succeed at node1, then it will succeed at the new leaseholder (either node2 or node3).
</blockquote></details>

Understood - @bdarnell also pointed this out. To clarify, though - this code is still correct, just not as optimistic as the previous code, right?

---

*[pkg/kv/dist_sender.go, line 1186 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1186:-KorjYtwxQpXL1NROvcW:bfbyd19) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1186)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

This comment should probably be adjusted to reflect the fact there may not actually be an error. It's confusing as is.
</blockquote></details>

Agreed; I've removed the comment altogether, I think the comment in the body of `if propagateError {` is sufficient.

---

*[pkg/kv/dist_sender.go, line 1220 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1220:-KorjdjbjviwU3ONFFPB:bubx5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1220)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Here, you're going to need to resurrect the `ambiguousError` variable and check `haveCommit` to properly return the `AmbiguousResultError` instead of whatever proximate error is received here.
</blockquote></details>

You mean if I resurrect the keep-trying-until-error logic, right?

---

*[pkg/kv/dist_sender.go, line 1227 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1227:-KorjonZYqOo7Yr85An3:b84b6kw) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1227)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Now that you've removed the `SendNextTimeout()`, nothing uses the `deadline` being set in the `grpcTransport`. This means that in the case where node2 and node3 believe node1 (AWOL) is still the leaseholder, node1 will keep getting moved to the front of the transport, will fail fast, then we'll try node2 and node3 in rapid succession, each putting node1 back into the front position.

The end result is that because you're no longer employing deadline to trigger a timer, this code will busy loop, sending RPCs to node2 and node3 until a Raft election and leaseholder transfer to node2 or node3.
</blockquote></details>

I don't follow this - the timer firing is only relevant in that it triggers calls to `transport.IsExhausted`, which is still happening. There's still a deadline on the clients, so assuming that the calls to nodes 2 and 3 complete faster than those deadlines we'll still exhaust the transport. Perhaps that deadline should be set higher, or perhaps I'm completely missing the point?

---

*[pkg/kv/transport.go, line 304 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KorNkbUmFIeOV82CsQW:-KorkYl6Mg5NldiW0bSS:b1otmko) ([raw file](https://github.com/cockroachdb/cockroach/blob/b238bb52542dc9c167b92f3567d960b5113e8d3a/pkg/kv/transport.go#L304)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

This unfortunately was an overlooked bug in our setting the `SendNextTimeout` value to 10 minutes. The case where it was incorrectly no longer used is as follows:

- Node1 is leaseholder and goes AWOL
- gRPC heartbeats indicate node1 is not reachable, and fail the RPC
- Node2 is tried and returns `Node1 is leaseholder`
- Node1 is put to front of transport
- Batch is sent to Node1 and fails fast
- Node3 is tried and returns `Node1 is leaseholder`
- Node1 is put to front of transport
- Batch is sent to Node1 and fails fast

Note that each time node2 and node3 fail with a `NotLeaseholderError`, they get the retryable flag set in the transport, which made them eligible for a retry. We do want to retry them because we expect either node2 or node3 to eventually secure the lease.
</blockquote></details>

In other words, we do want this value to be set reasonably low?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987)*

**Comment by spencerkimball** [comment]
Review status: 9 of 12 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/kv/dist_sender.go, line 1181 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1181:-KortZpq7GwQkHaPMGRA:b2dudgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1181)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Understood - @bdarnell also pointed this out. To clarify, though - this code is still correct, just not as optimistic as the previous code, right?
</blockquote></details>

Yes, it's still correct to return `AmbiguousResultError` early, but that means more confusion for users. My intuition is that in many or even most cases where a node goes missing, the original batch does not succeed and can succeed once another node becomes leaseholder.

---

*[pkg/kv/dist_sender.go, line 1220 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1220:-KortzF59PF6XwzsK2xD:b2nhswy) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1220)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

You mean if I resurrect the keep-trying-until-error logic, right?
</blockquote></details>

Yes

---

*[pkg/kv/dist_sender.go, line 1227 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1227:-KorvftGcdF3sWo2y7P9:b3wajj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1227)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I don't follow this - the timer firing is only relevant in that it triggers calls to `transport.IsExhausted`, which is still happening. There's still a deadline on the clients, so assuming that the calls to nodes 2 and 3 complete faster than those deadlines we'll still exhaust the transport. Perhaps that deadline should be set higher, or perhaps I'm completely missing the point?
</blockquote></details>

Answered in person.

---

*[pkg/kv/transport.go, line 304 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KorNkbUmFIeOV82CsQW:-Korvhd_UIX-WsFTiDMl:bmaifu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b238bb52542dc9c167b92f3567d960b5113e8d3a/pkg/kv/transport.go#L304)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

In other words, we do want this value to be set reasonably low?
</blockquote></details>

Yes, `1s` is reasonable, as set here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987)*

**Comment by tamird** [comment]
Review status: 0 of 16 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/kv/dist_sender.go, line 1181 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1181:-KpqeYuFsopBFPVHHpqf:b-4qtrg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1181)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Yes, it's still correct to return `AmbiguousResultError` early, but that means more confusion for users. My intuition is that in many or even most cases where a node goes missing, the original batch does not succeed and can succeed once another node becomes leaseholder.
</blockquote></details>

This should be fixed now, PTAL.

---

*[pkg/kv/dist_sender.go, line 1220 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1220:-Kpqec35xmQL7o94aTdc:bb71pro) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1220)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Yes
</blockquote></details>

PTAL

---

*[pkg/kv/dist_sender.go, line 1227 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1227:-Kpqee4dEp-3gt_C1AzH:bkfho4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1227)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Answered in person.
</blockquote></details>

I've removed the deadline on the clients entirely; I think you'll probably want to look at this again.

---

*[pkg/kv/transport.go, line 304 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KorNkbUmFIeOV82CsQW:-Kpqej_KkPLDH5NJdhtO:bxizgpu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b238bb52542dc9c167b92f3567d960b5113e8d3a/pkg/kv/transport.go#L304)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Yes, `1s` is reasonable, as set here.
</blockquote></details>

Removed :grimacing: 

---

*[pkg/sql/ambiguous_commit_test.go, line 50 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KorPr6Hc25G0qRIS3JT:-KpqemVFZ1-d_ID7zeAe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7177c64eb8cf74b41741185c0df6d5ac576ae5c6/pkg/sql/ambiguous_commit_test.go#L50)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Agreed; will work on restoring this test.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987)*

**Comment by tamird** [comment]
Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/kv/dist_sender.go, line 1227 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KomfkMBe2g77KD09J4A-r4-1227:-Kpqs9G978wbHGkz-1vP:b-uqn83) ([raw file](https://github.com/cockroachdb/cockroach/blob/bcefda1871854d77e564fdb46990360b835f0a21/pkg/kv/dist_sender.go#L1227)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I've removed the deadline on the clients entirely; I think you'll probably want to look at this again.
</blockquote></details>

Restored the 1-second deadline that was present in earlier iterations of this change.

---

*[pkg/kv/transport.go, line 304 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-KorNkbUmFIeOV82CsQW:-KpqsDk4Z9SngcnUW9cJ:b-31h4ti) ([raw file](https://github.com/cockroachdb/cockroach/blob/b238bb52542dc9c167b92f3567d960b5113e8d3a/pkg/kv/transport.go#L304)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Removed :grimacing: 
</blockquote></details>

Restored the 1-second deadline.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987)*

**Comment by bdarnell** [comment]
:lgtm:

---

Reviewed 18 of 18 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16987#-:-KpugxCYkb9XIiuR1Eay:bnfp4nl)*
